src/Subscription.hpp

```diff
@@ -0,0 +1,77 @@
+/**
+ * This software is distributed under the terms of the MIT License.
+ * Copyright (c) 2020 LXRobotics.
+ * Author: Alexander Entinger <alexander.entinger@lxrobotics.com>
+ * Contributors: https://github.com/107-systems/107-Arduino-Cyphal/graphs/contributors.
+ */
+
+#ifndef INC_107_ARDUINO_CYPHAL_SUBSCRIPTION_HPP
+#define INC_107_ARDUINO_CYPHAL_SUBSCRIPTION_HPP
+
+/**************************************************************************************
+ * INCLUDE
+ **************************************************************************************/
+
+#include <memory>
+
+/**************************************************************************************
+ * NAMESPACE
+ **************************************************************************************/
+
+namespace impl
+{
+
+/**************************************************************************************
+ * CLASS DECLARATION
+ **************************************************************************************/
+
+class SubscriptionBase
+{
+public:
+  virtual ~SubscriptionBase() { }
+  virtual void onTransferReceived(CanardRxTransfer const & transfer) = 0;
+};
+
+template <typename T>
+class Subscription : public SubscriptionBase
+{
+public:
+  Subscription(std::function<void(T const &)> on_receive_cb, std::function<void(void)> on_destruction_cb)
```

Comment:
You can cheaply avoid extra call indirection and heap allocation by replacing the std::functions here with native lambdas. Since you already have a type-erasing hierarchy here, this wouldn't affect the usage but potentially reduce the number of heap allocations per instance from 3 to 1.

Reply:
I'm not sure what you mean by native lambda. Could you provide an example :pray: ?

Reply:
Remove std::function and add more template arguments as shown below. The static asserts are optional.

```c++
template <typename T, typename OnReceiveCb, typename OnDestructionCb>
class Subscription final : public SubscriptionBase
{
  static_assert(std::is_invocable_v<OnReceiveCb, const (your callback arg type)&>);
  // <...>
public:
  Subscription(const OnReceiveCb& on_receive_cb, const OnDestructionCb& on_destruction_cb)
  : SubscriptionBase{}
  , _on_receive_cb{on_receive_cb}
  , _on_destruction_cb{on_destruction_cb}
  { }
  virtual ~Subscription();
  virtual void onTransferReceived(CanardRxTransfer const & transfer) override;
  [[nodiscard]] inline CanardRxSubscription & canard_rx_subscription() { return _canard_rx_sub; }
private:
  CanardRxSubscription _canard_rx_sub;
  OnReceiveCb     _on_receive_cb;
  OnDestructionCb _on_destruction_cb;
};
```

Reply:
Done in a537f374801d9b8d3ca661de6c799e7ba90bc9c4. Can you :pray: take a look if this implementation brings about the benefits that you expect it to have?

Reply:
Yes, looks good to me.